File: kombi-lib/src/main/java/com/sgnatiuk/combination/MapCombination.java
package com.sgnatiuk.combination;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;

class MapCombination<K, V> extends AbstractCombination<Map<K, V>> {

    private final Map<K, V> originData;

    protected MapCombination(Map<K, V> originData) {
        this(
                originData,
                new Range(1, CombinationKt.calculateCombinationsNumber(originData.size()))
        );
    }

    protected MapCombination(Map<K, V> originData, Range range) {
        super(range);
        this.originData = originData;
    }


    @Override
    Combination<Map<K, V>> subCombination(Range range) {
        return new MapCombination<>(originData, range);
    }

    @NotNull
    @Override
    public Iterator<Map<K, V>> iterator() {
        return new BinaryMaskIterator<>(
                range,
                new MapBuilder<>(
                        originData,
                        new ArrayList<>(originData.keySet())
                )
        );
    }
}

File: kombi-lib/src/main/kotlin/com/sgnatiuk/combination/MapCombination.kt
package com.sgnatiuk.combination

internal class MapCombination<K, V>(
        private val originData: Map<K, V>,
        range: LongRange = 1..originData.size.calculateCombinationsNumber()
) : AbstractCombination<Map<K, V>>(range) {

    override fun iterator() : Iterator<Map<K, V>> {
        return BinaryMaskIterator(
                range,
                MapBuilder(
                        originData,
                        originData.keys.toList()
                )
        )
    }

    override fun subCombination(range: LongRange): Combination<Map<K, V>> {
        return MapCombination(originData, range)
    }
}